tests/main/views/test_buyers.py

```diff
@@ -3700,3 +3737,110 @@ def test_flash_message_shown_on_brief_cancel(self):
         res = self.client.get(expected_url)
         flash_div = html.fromstring(res.get_data(as_text=True)).xpath('//div[@class="banner-success-without-action"]')
         assert flash_div[0].text_content().strip() == "You've updated 'My Amazing Brief'"
+
+
+class TestAwardOrCancelBrief(BaseApplicationTest):
+    url = '/buyers/frameworks/digital-outcomes-and-specialists-2/requirements/digital-outcomes/{brief_id}/award'
+
+    def setup_method(self, method):
+        super(TestAwardOrCancelBrief, self).setup_method(method)
+        self.data_api_client_patch = mock.patch('app.main.views.buyers.data_api_client', autospec=True)
+        self.data_api_client = self.data_api_client_patch.start()
+
+        self.data_api_client.get_framework.return_value = api_stubs.framework(
+            slug='digital-outcomes-and-specialists-2',
+            status='live',
+            lots=[
+                api_stubs.lot(slug='digital-outcomes', allows_brief=True),
+            ]
+        )
+        self.brief = api_stubs.brief(
+            user_id=123,
+            framework_slug='digital-outcomes-and-specialists-2',
+            lot_slug="digital-outcomes",
+            status='closed'
+        )['briefs']
+
+        self.data_api_client.get_brief.return_value = {"briefs": self.brief}
+        self.login_as_buyer()
+
+    def teardown_method(self, method):
+        self.data_api_client_patch.stop()
+        super(TestAwardOrCancelBrief, self).teardown_method(method)
+
+    def test_404_if_user_is_not_brief_owner(self):
+        self.data_api_client.get_brief.return_value['briefs']['users'][0]['id'] = 234
+
+        res = self.client.get(self.url.format(brief_id=self.brief["id"]))
+
+        assert res.status_code == 404
+
+    @pytest.mark.parametrize('status', ['withdrawn', 'draft', 'live', 'cancelled', 'unsuccessful', 'awarded'])
+    def test_404_if_brief_not_closed(self, status):
+        self.data_api_client.get_brief.return_value['briefs']['status'] = status
+
+        res = self.client.get(self.url.format(brief_id=self.brief["id"]))
+
+        assert res.status_code == 404
+
+    @pytest.mark.parametrize('framework_status', ['live', 'expired'])
+    def test_200_for_acceptable_framework_statuses(self, framework_status):
+        self.data_api_client.get_framework.return_value['frameworks']['status'] = framework_status
+
+        res = self.client.get(self.url.format(brief_id=123))
+
+        assert res.status_code == 200
+
+    def test_that_no_option_chosen_triggers_error(self):
+        res = self.client.post(self.url.format(brief_id=123))
+
+        document = html.fromstring(res.get_data(as_text=True))
+        validation_message = document.xpath('//span[@class="validation-message"]')[0].text_content()
+
+        assert res.status_code == 400
+        assert "You need to answer this question." in validation_message
+        assert self.data_api_client.cancel_brief.called is False
```

Comment:
I'm not 100% sure this assertion is useful? Do we ever cancel via the API in this view? (Also if it turns out we do need it, I'd rather have the assertion be `is None`!)

Reply:
Hmmm maybe I should move it ito another test called:
`test_that_no_option_chosen_does_not_trigger_update` ?

The comparison to False is a funny one. `.called` on a mock will always return either the [python `True` or python `False`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.called) object depending on whether the mock function was called. Because there are only ever one of each of these. and because the `is` keyword ensures that 2 objects have the same `id` `is` is the right comparison to use here.
```
In [1]: [] is []
Out[1]: False

In [1]: [] is False
Out[1]: False

In [2]: False is False
Out[2]: True
```

We can of course just use `assert not mock.called`. But because `0, (), [], {}, None, False, ''` are all `not True`, and because we know what we should be getting `is False` is the one.

On the other hand we can ignore all of the above because I should be using [`mock.assert_not_called()`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_not_called) apparently.

Reply:
😸 Yes I'd much prefer `assert_not_called` if we were going to keep it. Can we do the same for `test_400_if_incorrect_status_supplied`?

Reply:
Assert not called and broken out into different tests :+1: